### PR TITLE
NF: Remove hook

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This was discussed on https://www.google.com/url?q=https://groups.google.com/d/msgid/anki-android/64e5bb2c-36b4-442f-858c-37fdb9d6f927%2540googlegroups.com?utm_medium%3Demail%26utm_source%3Dfooter&source=gmail&ust=1590400772169000&usg=AFQjCNGNtVL2zpgVkV3rcxQHprw2mZ50rQ

I remove the hook package. I remove all "Hook" classes, and most classes which only contain a hook and a small function.

The reason to do so is:
* code is simpler to follow. For example, when processing a card content, instead of looking for all occurrence of "mungeQA", I just need to follow the functions dealing with chess and LaTeX.
* Code is smaller
* Code is well typed, as there are no more Object, and ...
* It will become far simpler to work differently for card content processing for browser and for reviewer. This will allow to easily show LaTeX (and chess notation ?) inside of browser, instead of deleting it as we currently do.

I also moved the remaining classes (advanced stats and chess) to other package where it makes more sens.

As usual, the change is "big" (more than 500 lines) and my seem complex. To make the task easier, I split it into a lot of small commit. It should be easy to check that none of those commit change the behavior of AnkiDroid in any way; and so the full PR does not change the behavior either.

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
 
 import timber.log.Timber;
 
-public class ChessFilter extends Hook {
+public class ChessFilter {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java:27]: Codacy Issue found: [The utility class name 'ChessFilter' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 2
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  41
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/template/FuriganaFilters.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Hint.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java  8
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/AdvancedStatistics.java  16
         

Clones added
============
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687&bid=18279834)

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Hint.java [hunk 2]
```diff
@@ -14,32 +14,24 @@
...
-                return "";
-            }
-            Resources res = AnkiDroidApp.getAppResources();
-            // random id
-            String domid = "hint" + txt.hashCode();
-            return "<a class=hint href=\"#\" onclick=\"this.style.display='none';document.getElementById('" +
-                    domid + "').style.display='block';_relinquishFocus();return false;\">" +
-                    res.getString(R.string.show_hint, (String) args[2]) + "</a><div id=\"" +
-                    domid + "\" class=hint style=\"display: none\">" + txt + "</div>";
+public class Hint {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/Hint.java:24]: Codacy Issue found: [The utility class name 'Hint' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

**Comment by Arthur-Milchior** [comment]
Pushed again to correct the error 
"Codacy Issue found: The utility class name 'Hint' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'"

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 6
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/template/FuriganaFilters.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  41
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/AdvancedStatistics.java  16
- AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java  8
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Hint.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java  1
         

Clones added
============
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  7
- AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687&bid=18279834)

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java (hunk 1 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/ChessFilter.java:27]: Codacy Issue found: [The utility class name 'ChessFilter' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 3]
```diff
@@ -2243,4 +2243,71 @@ public boolean getFailed() {
             return mFailed;
         }
     }
+
+    private static final Pattern fFenPattern = Pattern.compile("\\[fen ?([^\\]]*)\\]([^\\[]+)\\[/fen\\]");
+    private static final Pattern fFenOrientationPattern = Pattern.compile("orientation *= *\"?(black|white)\"?");
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2248]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 4]
```diff
@@ -2243,4 +2243,71 @@ public boolean getFailed() {
...
+                    "    fentxt=fentxt.replace(/P/g,'<td>&#9817;</td>');" +
+                    "    fentxt=fentxt.replace(/k/g,'<td>&#9818;</td>');" +
+                    "    fentxt=fentxt.replace(/q/g,'<td>&#9819;</td>');" +
+                    "    fentxt=fentxt.replace(/x/g,'<td>&#9820;</td>');" +
+                    "    fentxt=fentxt.replace(/b/g,'<td>&#9821;</td>');" +
+                    "    fentxt=fentxt.replace(/n/g,'<td>&#9822;</td>');" +
+                    "    fentxt=fentxt.replace(/p/g,'<td>&#9823;</td>');" +
+                    "    return '<div align=\"center\" width=\"100%%\"><table class=\"chess_board\" cellspacing=\"0\" cellpadding=\"0\"><tr>'+fentxt+'</tr></table></div>';" +
+                    "})('%s', %b)";
+    private static boolean sChessEnabled = false;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2279]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 5]
```diff
@@ -2243,4 +2243,71 @@ public boolean getFailed() {
             return mFailed;
         }
     }
+
+    private static final Pattern fFenPattern = Pattern.compile("\\[fen ?([^\\]]*)\\]([^\\[]+)\\[/fen\\]");
+    private static final Pattern fFenOrientationPattern = Pattern.compile("orientation *= *\"?(black|white)\"?");
+    private static final String fRenderFen =
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2249]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 6]
```diff
@@ -2243,4 +2243,71 @@ public boolean getFailed() {
             return mFailed;
         }
     }
+
+    private static final Pattern fFenPattern = Pattern.compile("\\[fen ?([^\\]]*)\\]([^\\[]+)\\[/fen\\]");
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2247]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Hint.java (hunk 2 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/Hint.java:24]: Codacy Issue found: [The utility class name 'Hint' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

**Comment by david-allison-1** [comment]
I'm finding this hard to review due to the size, if it's easily possible, could this be split across one pull request per hook removed?

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 4
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/template/FuriganaFilters.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  41
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/AdvancedStatistics.java  16
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java  1
         

Clones added
============
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687&bid=18279834)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 5 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2249]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 3 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2248]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 4 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2279]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 6 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2247]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 7]
```diff
@@ -2701,4 +2702,25 @@ public void setReportLimit(int reportLimit) {
     }
 
     /** End #5666 */
+
+    protected static void leech(Card card, Activity activity) {
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:2706]: Yes

**Comment by david-allison-1** [comment]
My main take from this is that we're moving from a singleton to more spread out global, mutable state, which makes Anki harder to reason about and test. 

This is a "we should step back and ask whether there are other designs that we can take". I'd be happy to see the code removed, but I'd like to ensure that we've considered other options.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java]: Done.

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 6
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  41
- AnkiDroid/src/main/java/com/ichi2/libanki/template/FuriganaFilters.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/AdvancedStatistics.java  16
         

Clones added
============
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687&bid=18279834)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 3 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2248]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 8]
```diff
@@ -22,13 +22,15 @@
 import android.app.Activity;
 
 import android.content.Context;
+import android.content.res.Resources;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:25]: Codacy Issue found: [Avoid unused imports such as 'android.content.res.Resources'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 4 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2279]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java [hunk 9]
```diff
@@ -22,13 +22,15 @@
...
 import android.database.Cursor;
 import android.database.SQLException;
 import android.database.sqlite.SQLiteConstraintException;
 import android.graphics.Typeface;
 import android.text.SpannableStringBuilder;
 import android.text.TextUtils;
 import android.text.style.StyleSpan;
+import android.widget.Toast;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:33]: Codacy Issue found: [Avoid unused imports such as 'android.widget.Toast'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 5 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2249]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 6 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:2247]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

**Comment by Arthur-Milchior** [comment]
Please note that I can't push anymore PR related to hooks until the other PR are merged. Because the hint and Furigana filters can't easily be removed until the other runFilter are removed

**Comment by Arthur-Milchior** [comment]
> My main take from this is that we're moving from a singleton to more spread out global, mutable state, which makes Anki harder to reason about and test.
> 
> This is a "we should step back and ask whether there are other designs that we can take". I'd be happy to see the code removed, but I'd like to ensure that we've considered other options.

Which singleton are you referring to ?
The only singleton that were here was Hook. It is not here anymore because there is no need for hook manager.
There are only two states: should we interpreted chess notation and should we use advanced statistics. Instead of installing/uninstalling hooks, we simply have a boolean near the function to recall the value in the preference. Actually, we could even remove the boolean and check the preference directly in the function. I never really understood how to access preference if I don't already have either them or a Context.

Appart from that, I replaced indirection by direct call. For example, instead of saying that you want to call the hook for hint, you directly call the method dealing with hints. So it seems to actually be less "spread out" since there is not a single function "runHook" which is in charge of doing multiple unrelated things

**Comment by timrae** [comment]
I vote against this change.

Removing hooks is fine if you replace them with something like Mike was talking about on that thread, but integrating the code into libanki is going in the opposite direction from where we want to be IMO.

The main AnkiDroid codebase and certainly the libanki package should be focused as much as possible on features that are in Anki desktop. As much as possible we want other features to be implemented independently of AnkiDroid as a separate code base. The plan with hooks was always to eventually achieve this goal, and there are several possibilities for having a plugin like system for AnkiDroid, it's just nobody has put the effort in to fully achieve this yet.

**Comment by Arthur-Milchior** [comment]
The way I did understand Mike remark was that the feature implemented by the hooks should remain (i.e. not loosing hint, latex, chess, advanced stats, leech and Furigana), but that any feature that is not directly coded into AnkiDroid would be dealt in the webview, thanks to js/ts in card template; i.e. that those particular hooks would not be used. This is why I went with those change.

**Comment by Arthur-Milchior** [comment]
For the sake of clarity, this was rebased as follow:
* First I merged all PR done independently
* Then I rebased over this merge
The goal are:
* It would incorporate all correction made in other PR
* it shows the result when you merge everything, what is the expected final result if you merge those commits
* the last commits are the one that will be useful to create more PR when it become possible

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 5
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/template/FuriganaFilters.java  2
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  41
- AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/AdvancedStatistics.java  16
         

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/libanki/hooks/Hooks.java  -4
         

Clones added
============
- AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687&bid=18279834)

File: AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java [hunk 10]
```diff
@@ -18,13 +18,16 @@
 
 
 
+import android.content.SharedPreferences;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java:21]: Codacy Issue found: [Avoid unused imports such as 'android.content.SharedPreferences'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java [hunk 11]
```diff
@@ -18,13 +18,16 @@
...
+import android.preference.PreferenceManager;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java:22]: Codacy Issue found: [Avoid unused imports such as 'android.preference.PreferenceManager'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java [hunk 12]
```diff
@@ -18,13 +18,16 @@
...
+
 import java.util.Locale;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 import timber.log.Timber;
 
-public class ChessFilter extends Hook {
+public class ChessFilter {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java:30]: Codacy Issue found: [The utility class name 'ChessFilter' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 9 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:33]: Codacy Issue found: [Avoid unused imports such as 'android.widget.Toast'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java (hunk 8 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:25]: Codacy Issue found: [Avoid unused imports such as 'android.content.res.Resources'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

**Comment by mikehardy** [comment]
> I vote against this change.
> 
> Removing hooks is fine if you replace them with something like Mike was talking about on that thread, but integrating the code into libanki is going in the opposite direction from where we want to be IMO.
> 
> The main AnkiDroid codebase and certainly the libanki package should be focused as much as possible on features that are in Anki desktop. As much as possible we want other features to be implemented independently of AnkiDroid as a separate code base. The plan with hooks was always to eventually achieve this goal, and there are several possibilities for having a plugin like system for AnkiDroid, it's just nobody has put the effort in to fully achieve this yet.

@timrae I'm not sure I'm against the change - I need to review - but I'm tentatively for it at the same time I do agree (as is visible in my comments about possible hook future) we should have a plugin system somewhere. I think it just wasn't possible before as web tech was missing at project inception and platform restrictions meant without the dynamic capabilities of web tech plugins had to be bottled up in the main tree (a closed "open" plugin ecosystem is neither, apparently).

But I think it's generally agreed amongst Damien and everyone here that libanki is Rust in the future and AnkiDroid will wrap it and JNI it for use by the UI, which is a future state that is exactly as you describe - AnkiDroid focusing on Desktop functionality and a vanilla libanki. I know Arthur is aware of that, David has a hello-world Rust/JNI test cooking already and is probing what it "feels" like (does it allow debugging into libanki? etc). 

So the question for me is does this make that future change worse or better or is it just sideways motion? I don't mind boiling down the generic Hooks system without a replacement here personally, because it was empirically pointless (c.f. change frequency, new feature frequency) but if it makes libanki more AnkiDroid-custom then I don't like it. I'll see that in review

**Comment by timrae** [comment]
OK I just took another quick look at the diff and it doesn't seem as bad as I initially felt. Does Rust libanki have any hook architecture?

**Comment by mikehardy** [comment]
> OK I just took another quick look at the diff and it doesn't seem as bad (in terms of plugin specific code in libanki package) as I initially felt. Does Rust libanki have any hook architecture?

@timrae tagging in @Arthur-Milchior I believe he may be our closest tracker of libanki changes, and if it has a plugin architecture he may know. I personally have been fully utilized w/review+merge+release here and haven't looked up to scan libanki recently, trusting in the general direction + that David is probing rust integration

**Comment by Arthur-Milchior** [comment]
As far as I know, there are no hooks directly in Rust. All hooks are either applied before rust is called or after it is called, to pre/post-process the results. This mean that some hooks are actually changed/moved. They used to be call during a computationally costly process, sometime many time in a loop. Now the costly process is done in rust, and so hooks had to be put before/after.

There is at least one-specific case which is pretty relevant here.
If you have ``{{feature:field_name}}`` in your template, as you may know, some process is applied. If feature is either hint or Furigana related, then we process it. Otherwise ankidroid just ignore it. In Anki, it looks for a hook "fmod_feature" and apply it. Since hooks are only in Python, what actually occurs is that rust processing halts and rust let python ends the rendering of the card content. This is slower, but I do not believe there was other solution as rust would not easily allows add-on as anki.

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687&bid=18279834)

File: AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java [hunk 13]
```diff
@@ -28,7 +28,7 @@
...
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java:31]: Codacy Issue found: [The utility class name 'ChessFilter' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5575687)

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/hooks/ChessFilter.java:31]: Don't hesitate to give another name

**Comment by Arthur-Milchior** [comment]
Here it is. No more hooks !

File: AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java [hunk 14]
```diff
@@ -46,7 +45,6 @@ public static Collection Collection(Context context, String path) {
     public static Collection Collection(Context context, String path, boolean server, boolean log) {
         assert path.endsWith(".anki2");
         // Since this is the entry point into libanki, initialize the hooks here.
```
Comment (Kubo2) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:47]: Looks like forgot to remove?

**Comment by Arthur-Milchior** [comment]
Good catch. You want to do a pr to remoce it ?

**Pull request merged**